core/src/main/java/com/linecorp/armeria/internal/common/util/SslContextUtil.java

```diff
@@ -241,6 +229,132 @@ private static String badCipherSuiteMessage(String cipher) {
+        if (keyPair != null) {
+            builder.keyManager(keyPair.privateKey(), keyPair.certificateChain());
+        }
+        builder.endpointIdentificationAlgorithm("HTTPS");
+
+        applyCommonConfigs(clientTlsSpec, builder);
+
+        return builder.build();
+    }
+
+    private static List<String> filterProtocols(Collection<String> protocols, SslProvider provider) {
+        final Set<String> supportedProtocols = supportedProtocols(provider);
+        final List<String> filtered = protocols.stream()
+                                               .filter(supportedProtocols::contains)
+                                               .collect(toImmutableList());
+        if (filtered.isEmpty()) {
+            throw new IllegalStateException(provider + " supports none of " + protocols);
+        }
+
+        if (!warnedUnsupportedProtocols && protocols.size() != filtered.size()) {
+            warnedUnsupportedProtocols = true;
+            if (logger.isDebugEnabled()) {
+                final List<String> missingProtocols = protocols.stream()
+                                                               .filter(p -> !filtered.contains(p))
+                                                               .collect(toImmutableList());
+                logger.debug("{} does not support: {}", provider, missingProtocols);
+            }
+        }
+        return filtered;
+    }
+
+    private static void validateSslContext(boolean tlsAllowUnsafeCiphers, SslContext sslContext) {
+        final Set<String> ciphers = ImmutableSet.copyOf(sslContext.cipherSuites());
+        checkState(!ciphers.isEmpty(),
+                   "SSLContext has no cipher suites enabled. " +
+                   "You must specify at least one cipher suite.");
+        if (sslContext.applicationProtocolNegotiator().protocols().contains(ApplicationProtocolNames.HTTP_2)) {
+            validateHttp2Ciphers(ciphers, tlsAllowUnsafeCiphers);
+        }
+    }
+
+    private static void applyCommonConfigs(AbstractTlsSpec tlsSpec, SslContextBuilder builder)
+            throws Exception {
+        if (tlsSpec.verifierFactories().isEmpty()) {
+            if (!tlsSpec.trustedCertificates().isEmpty()) {
+                builder.trustManager(tlsSpec.trustedCertificates());
+            }
+        } else {
+            final KeyStore ks = toKeyStore(tlsSpec.trustedCertificates());
+            final X509ExtendedTrustManager delegate = defaultPkixTrustManager(ks);
+            builder.trustManager(toTrustManager(tlsSpec, delegate));
+        }
+
+        final List<String> protocols = filterProtocols(tlsSpec.protocols(),
+                                                       tlsSpec.engineType().sslProvider());
+        builder.protocols(protocols);
+        builder.ciphers(tlsSpec.ciphers(), SupportedCipherSuiteFilter.INSTANCE);
+        builder.sslProvider(tlsSpec.engineType().sslProvider());
+
+        tlsSpec.tlsCustomizer().accept(builder);
```

Comment:
Which method are you going to use instead of `tlsCustomizer` after it's removed?